### PR TITLE
Don't duplicate ParamSpec prefixes and properly substitute Paramspecs

### LINKED ISSUE #12909: False postives when decorating functions with signature that uses ParamSpec within a decoarator
**Bug Report**

When using a decorator on the decorated function within the definition of another decorator, mypy reports false positives.

**To Reproduce**

```python3
from typing import Callable, TypeVar
from typing_extensions import ParamSpec, Concatenate

P = ParamSpec("P")
T = TypeVar("T")


def identity(func: Callable[P, T]) -> Callable[P, T]:
    def _wrapped(*args: P.args, **kwargs: P.kwargs) -> T:
        return func(*args, **kwargs)

    return _wrapped


def int_identity(func: Callable[Concatenate[int, P], T]) -> Callable[Concatenate[int, P], T]:
    # identify should preserve the signature of _wrapped
    # error: Argument 1 to "identity" has incompatible type "Callable[[Arg(int, 'number'), **P], T]"; expected "Callable[[Arg(int, 'number'), **P], T]"
    @identity
    def _wrapped(number: int, *args: P.args, **kwargs: P.kwargs) -> T:
        return func(number, *args, **kwargs)

    return _wrapped
```

**Expected Behavior**

The code should pass type-check with no errors.

**Actual Behavior**

Got error:
`error: Argument 1 to "identity" has incompatible type "Callable[[Arg(int, 'number'), **P], T]"; expected "Callable[[Arg(int, 'number'), 
**P], T]"`

**Your Environment**

- Mypy version used: 0.960
- Mypy command-line flags: `--strict`
- Python version used: Python 3.8.10
- Operating system and version: WSL2 - Ubuntu 20.04

### PR DESCRIPTION
**Pull request by A5rocks**
I ran into the duplication issue personally, but this also fixes https://github.com/python/mypy/issues/12734

Also:
 - Fixes #12909

### CODE REVIEW

**Comment by A5rocks** [comment]
Minimum reproducer for failure:
```py
from typing import Callable, Any, overload
from typing_extensions import Concatenate, ParamSpec

P = ParamSpec("P")

@overload
def command() -> Callable[[Callable[Concatenate[object, object, P], object]], None]:
    ...

@overload
def command(
    cls: int = ...,
) -> Callable[[Callable[Concatenate[object, P], object]], None]:
    ...

def command(
    cls: int = 42,
) -> Any:
    ...
```

hmmmmm

**Comment by A5rocks** [comment]
FYI, while looking through issues to see if this PR fixes anything else, I found that https://github.com/python/mypy/issues/14169 is untagged (it should have the `topic-paramspec` tag)

File: mypy/expandtype.py [hunk 1]
```diff
@@ -370,6 +373,23 @@ def visit_callable_type(self, t: CallableType) -> Type:
...
+            #       be done once; we should update that "once" location rather than here.
+            #       (see testAlreadyExpandedCallableWithParamSpecReplacement)
+            elif isinstance(repl, ParamSpecType) and len(t.arg_types) == 2:
+                # We're substituting one paramspec for another; this can mean that the prefix
+                # changes. (e.g. sub Concatenate[int, P] for Q)
+                prefix = repl.prefix
+                old_prefix = param_spec.prefix
+
+                # Check assumptions. I'm not sure what order to switch these:
+                assert not old_prefix.arg_types or not prefix.arg_types
```
Comment (JukkaL) [inline mypy/expandtype.py:386]: This looks a bit suspicious to me. Why can't both the arg types truthy?

**Review by JukkaL — COMMENTED** [review]
Thanks for the PR! I have a few questions.

File: mypy/types.py [hunk 2]
```diff
@@ -1970,22 +1970,19 @@ def param_spec(self) -> ParamSpecType | None:
         if self.arg_kinds[-2] != ARG_STAR or self.arg_kinds[-1] != ARG_STAR2:
             return None
         arg_type = self.arg_types[-2]
-        if not isinstance(arg_type, ParamSpecType):
+        if not isinstance(arg_type, ParamSpecType) or not isinstance(
+            self.arg_types[-1], ParamSpecType
+        ):
```
[the remark this answers is not in the document]
Comment (A5rocks) [reply mypy/types.py:1975]: Defensive, I can remove it.

File: mypy/expandtype.py (hunk 1 above)
Comment (A5rocks) [reply mypy/expandtype.py:386]: I haven't found a case where they were both truthy and yet not equal to each other yet -- I don't know what order to mix their arguments together. My gut says what I wrote next is correct but I would rather have an explicit test case (or be shown that this never happens)

File: mypy/expandtype.py [hunk 3]
```diff
@@ -370,6 +373,23 @@ def visit_callable_type(self, t: CallableType) -> Type:
                     ret_type=t.ret_type.accept(self),
                     type_guard=(t.type_guard.accept(self) if t.type_guard is not None else None),
                 )
+            # TODO: it seems this only has to be done *sometimes*. Conceptually this should only
+            #       be done once; we should update that "once" location rather than here.
+            #       (see testAlreadyExpandedCallableWithParamSpecReplacement)
```
[the remark this answers is not in the document]
Comment (A5rocks) [reply mypy/expandtype.py:378]: Hm, probably. I'm essentially saying "this is a band-aid; try this test without the arg count check to try investigating more"

I could probably do this todo now. I'll give it a shot...

[the remark this answers is not in the document]
Comment (A5rocks) [reply mypy/expandtype.py:378]: I've looked at this some more and I can't figure it out, I've tried my hand at making it more understandable.

**Comment by A5rocks** [comment]
Just a note that this exposes a bug in the constraint finder: I've got a local branch that fixes that.

PS This is ready for another review :D

**Comment by A5rocks** [comment]
I have a question about implementation and I'm not sure where to ask it (is there a better spot?), so here goes! Hopefully people who know the implementation of mypy check here :P

Essentially, the goal of this PR (and the followup PR which ATM is just over at https://github.com/A5rocks/mypy/tree/fix-fundamental-paramspecs) is to get this following code to typecheck:
```py
from typing import ParamSpec, Concatenate, Generic, TypeVar, Any
from collections.abc import Callable

P = ParamSpec("P")
T = TypeVar("T")
V = TypeVar("V")
R = TypeVar("R")

# ----- INSIDE BLOOM -----

def command_builder() -> Callable[[Callable[Concatenate[T, P], R]], Callable[P, Callable[[T], R]]]:
    def transformer(f: Callable[Concatenate[T, P], R], /) -> Callable[P, Callable[[T], R]]:
        def returned(*args: P.args, **kwargs: P.kwargs) -> Callable[[T], R]:
            def returned_transformer(z: T) -> R:
                return f(z, *args, **kwargs)

            return returned_transformer
        
        return returned

    return transformer

# ----- TEST CASES -----
class Example(Generic[P]):
    pass

@command_builder()
def test1(ex: Example[P], /) -> Example[Concatenate[int, P]]:
    ...

reveal_type(test1)  # Is OK... But def () -> def [P] (repro.Example[P`-1]) -> repro.Example[[int, **P`-1]]
reveal_type(test1())  # Should be def [P](?) (repro.Example[P`-1]) -> repro.Example[[int, **P`-1]]
```

ATM on that branch mypy spits out:
```
repro.py:28: error: Missing return statement  [empty-body]
repro.py:31: note: Revealed type is "def [P] () -> def (repro.Example[P`-1]) -> repro.Example[[builtins.int, **P`-1]]"
repro.py:32: note: Revealed type is "def (repro.Example[<nothing>]) -> repro.Example[<nothing>]"
Found 1 error in 1 file (checked 1 source file)
```

Obviously that last thing is *wrong*. (It has bad consequences for actual usage but even just the revealed type is wrong).

The problem here is that I am calling a "def [P] () -> def (repro.Example[P\`-1]) -> repro.Example[[builtins.int, **P\`-1]]", which in turn tries to infer `P`. There are no constraints as there are no arguments, meaning that `P` becomes `UninhabitedType` which is `<nothing>`. I'm not sure how to solve this. My preferred outcome is that this turns into "def [P] (repro.Example[P\`-1] -> repro.Example[[builtins.int, **P\`-1]]".

Basically, how should I solve this? My main idea revolves about making solving for constraints check if that type var appears in the arguments and if it doesn't then it gets "assigned" to the return type (if that's possible, which is only with stuff like callables). However, I'm not sure how to even check whether a type var appears in the arguments, let alone if there's a better way to approach this.

**Comment by A5rocks** [comment]
Hello again! Another review would be much appreciated -- the followup PR fixes a few open issues and I can't open that until this is merged (as far as I know?) without messing up commit history for myself (?).

**Review by hauntsaninja — APPROVED** [review]
Thanks, this looks good to me and am excited to see your future ParamSpec improvements. I'll wait a day or two in case Jukka wants to take another look; if I don't merge soon, just ping me :-)

**Comment by hauntsaninja** [comment]
Thank you!
